### ISSUE TITLE
Loading/Pushing a frontViewController during app startup issues - race with Storyboard?

### ISSUE THREAD

**Issue by bensebborn** [John-Lluch/SWRevealViewController#584]
Hi

I have an app which uses storyboards, from which the SWRevelViewController is setup with a side menu, so this loads automatically upon launch.

By default CategoriesViewController loads as the front ViewController.

I'm then trying to handle push notifications, using the code below within appdelegate. This should load a new VC (EventViewController) into the navigation stack so it shows.

`  
    /\* Using storyboards, so get main storyboard and reference VC from there */

```
 UIStoryboard *mainStoryboard = [UIStoryboard storyboardWithName:@"Main" bundle: nil];

    EventViewController *vc = (EventViewController*)[mainStoryboard instantiateViewControllerWithIdentifier:@"EventVC"];
    CategoriesViewController *catsVC = [mainStoryboard instantiateViewControllerWithIdentifier:@"CatsVC"];

    vc.ilid = ilid;

    SWRevealViewController *controller = (SWRevealViewController*)self.window.rootViewController;

    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:catsVC];
    [navController setViewControllers:@[catsVC,vc] animated:NO];

    [controller pushFrontViewController:navController animated:NO];
```

`

The idea here is that our main (catsVC) controller is loaded, then the eventsViewController (vc) is loaded, so we can navigate back correctly if we wish.

This works great if the above code is triggered while the app is in the foreground or background. 

However, if we run the above code within application didFinishLaunchingWithOptions  (ie responding to a push notification when the app is not running anywhere), the EventViewController never shows.

There are no errors so I am wondering if there is a race condition between the above code, and the storyboard, so the above loads first, then the storyboard is loading the CategoriesViewController so we can't see what's happening?

Any tips on getting the above code working for all scenarios (foreground, background and not running)

Many thanks

**Comment by bensebborn** [comment]
I have confirmed this is a race condition, because if I add a delay around the pushFrontViewController code as follows:

```
  int64_t delayInSeconds = 3.0;
      dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delayInSeconds * NSEC_PER_SEC);
      dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
           NSLog(@"Loading vc now ");
           [controller pushFrontViewController:navController animated:NO];
              });

        }
```

This works fine. So, how can I make this work without the forced delay?

**Comment by quetool** [comment]
Hi, i used you code in this way in appDelegate

```
UIStoryboard *main = [UIStoryboard storyboardWithName:@"Main" bundle: nil];

MyViewController *viewController = (MyViewController *)[main instantiateViewControllerWithIdentifier:@"ComentariosViewController"];

SWRevealViewController *controller = (SWRevealViewController *)self.window.rootViewController;

UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:viewController];
[navController setViewControllers:@[viewController] animated:YES];

[controller pushFrontViewController:navController animated:YES];
```

And it works for me, but i can not go back. Do you know how to do? How call popViewController?

**Comment by iDevelopper** [comment]
There is only one controller in the navigation controller's stack! Why do you want to pop?

**Comment by quetool** [comment]
i'll try to explain: when i receive a push notification, i want to open (push in root navigation controller) a particular view controller, then, i want to go back to de previous opened view controller.

**Comment by iDevelopper** [comment]
When you receive a notification, you push a controller with SW. Do the same when you want to open again the previous one:

In your controller, when you finish the job:
[self.revealViewController pushFrontViewController: previousController animated:YES];

**Comment by quetool** [comment]
but the thing is the view controller i open with the notification is not a reveal view controller

This is my main view controller (a reveal view controller type)
![1](https://cloud.githubusercontent.com/assets/14978705/15808189/c9bd7fbc-2b47-11e6-94bd-6421ff55d3b9.jpg)

This is the second view controller, the one i open with the notification, is not a reveal view controller. (just a normal view controller)
![2](https://cloud.githubusercontent.com/assets/14978705/15808190/d2bbd014-2b47-11e6-923d-d2d3360772f3.jpg)

When i open the 2nd view controller with the notification, i can not go back to the first view controller. That is my problem.

In fact, the backButtonItem is not being show after i opened by the notification.

This is the way i open the 2nd view controller from app delegate, may be im doing something wrong:

```
UIStoryboard *main = [UIStoryboard storyboardWithName:@"Main" bundle: nil];

SecondViewController *viewController = (SecondViewController *)[main instantiateViewControllerWithIdentifier:@"SecondViewController"];

SWRevealViewController *controller = (SWRevealViewController *)self.window.rootViewController;

UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:viewController];
[navController setViewControllers:@[viewController] animated:YES];

[controller pushFrontViewController:navController animated:YES];
```

**Comment by iDevelopper** [comment]
Yes you push the second view controller with SW then this second view controller has a property revealViewController. When you want to open the main view controller, just do:

[self.revealViewController pushFrontViewController: mainController animated:YES];

**Comment by quetool** [comment]
Ok, i'm not sure if that is the behavior i want. Is not logic with the behavior of my app. I need something like popViewController to return to the previous view controller opened by the user. I understand with the method you describe me, mainController it has to be a fixed view controller defined by me. That is what i do not want. 

But, trying something like passing in mainController the topViewController like this

```
- (UIViewController *) topMostController {
    UIViewController *topController = [UIApplication sharedApplication].keyWindow.rootViewController;
    while (topController.presentedViewController) {
        topController = topController.presentedViewController;
    }
    return topController;
}
```

Then calling it by your method

`[self.revealViewController pushFrontViewController:topController animated:YES];`

Im getting an uncaught exception.

`Uncaught exception: adding a root view controller <SWRevealViewController: 0x14d022a00> as a child of view controller:<SWRevealViewController: 0x14d022a00>`

Sorry for my english, i hope you understand!

**Comment by iDevelopper** [comment]
What is the class of topController? Use NSLog

**Comment by quetool** [comment]
Yes, i don't know why but the class is always (and is the reason why the error) SWRevealViewController

**Comment by iDevelopper** [comment]
Yes it is normal. You want to show the las view controller before the user has received the notification?

**Comment by quetool** [comment]
Yes, that is what i want. But i am coming to realize that is not possible without some "workaround", right?

**Comment by quetool** [comment]
like i said, i "worked around". Instead of open the 2nd view controller by the push, i open the 1st, and do a segue to the 2nd by passing all i need. In this way i can go back normaly with the backbuttonitem. Thanks for the help!

Is not the best solution but is close enough

**Comment by quetool** [comment]
Is not the best solución bit is close enough

**Comment by iDevelopper** [comment]
Si, I think it's a good solution. I not understood the question very well at the beginning. I thought when you received a notification you wanted to display a front view controller and then you wanted to display the front view controller on which was the user before. It is not a worked around. Well done!

**Comment by iDevelopper** [comment]
You could just add a test to know if the user is already in this first view controller and if it is the case just push the second view controller with self.navigationController

**Comment by quetool** [comment]
yes but i don't know how, since i always get SWRevealViewController as a class of all view controllers... plus i don't know how to identifier every view controller because all identifiers appears null.

**Comment by iDevelopper** [comment]
You can use this (see the tutorial : https://www.raywenderlich.com/117471/state-restoration-tutorial) or you can check the class of revealViewController.frontViewController.topViewController (with cast)

**Comment by quetool** [comment]
how do implement the 2nd option in appDelegate?
Because i tried:

`SWRevealViewController *controller = (SWRevealViewController *)self.window.rootViewController;`

Then

`controller.frontViewController.topViewController`

But topViewController does not exists in frontViewController

**Comment by iDevelopper** [comment]
SWRevealViewController *revealController = (SWRevealViewController *)self.window.rootViewController;
UINavigationController *nc = (UINavigationController *) revealController.frontViewController;
UIViewController *topController = nc.topViewController;
if ([topController isKindOfClass: [YourClass class]])
   ....

I am withy computer but it is something like this!

**Comment by quetool** [comment]
Yes, that is right and helpfull! thanks very much!

**Comment by bintu1234** [comment]
@quetool  HI I am also facing same issue.I see above all messages but I am unable to achieve that. I am also same scenario what you have faced.Finally what you did please help me

**Comment by quetool** [comment]
@bintu1234 what is your problem exactly?

**Comment by bintu1234** [comment]
@quetool  Thanks for the replay.I am using SWRevealViewController every thing is working fine.Now when i tap on Local notification i want to redirect to specific page. When i use below code in `didReceiveLocalNotification` it is working but there is no back button.
`UIStoryboard *main = [UIStoryboard storyboardWithName:@"Main" bundle: nil];
        ASNMessageChat *viewController = (ASNMessageChat *)[main instantiateViewControllerWithIdentifier:@"ChatView"];
        SWRevealViewController *controller = (SWRevealViewController *)self.window.rootViewController;
        UINavigationController *navController1 = [[UINavigationController alloc] initWithRootViewController:viewController];
        [navController1 setViewControllers:@[viewController] animated:YES];
       [controller pushFrontViewController: navController1 animated:YES];`
And one more thing is My destination is `ASNMessageChat` But it is not the Rear or Front for `SWRevealViewController` same as you explained in Screen shots above.  How to achieve this. Thank you.

**Comment by quetool** [comment]
What i did was a "workaround". 
With the notification I open my main view controller (just like you do) at the same time i set a BOOL property of that view controller to YES. 
If this property is setted to YES (because im coming from the notification), what i do is a segue to the view controller i wanted in the first place.

Now, this is not the behavior i was searching, but for the moment it works for me. I have planned to find a better solution in the future.

**Comment by bintu1234** [comment]
Ok.If you find better solution Please try to notify me.Thank you
